scilpy/denoise/asym_filtering.py

```diff
@@ -1,387 +1,480 @@
+        dot = directions.dot(directions.T)
+        x = np.arccos(np.clip(dot, -1.0, 1.0))
+        weights = _evaluate_gaussian_distribution(x, sigma_angle)
+        mask = x > (3.0*sigma_angle)
+        weights[mask] = 0.0
+        weights /= np.sum(weights, axis=-1)
+    else:
+        weights = np.eye(len(directions))
+    return weights
 
 
-def _evaluate_gaussian_distribution(x, sigma):
+def _unified_filter_build_nx(filter_shape, sigma_spatial, sigma_align,
+                             sphere, exclude_center):
     """
-    1-dimensional 0-centered Gaussian distribution
-    with standard deviation sigma.
+    Build the combined spatial and alignment filter.
 
     Parameters
     ----------
-    x: ndarray or float
-        Points where the distribution is evaluated.
-    sigma: float
-        Standard deviation.
+    filter_shape: tuple
+        Dimensions of filtering window.
+    sigma_spatial: float or None
+        Standard deviation of spatial filter. None disables Gaussian
+        weighting for spatial filtering.
+    sigma_align: float or None
+        Standard deviation of the alignment filter. None disables Gaussian
+        weighting for alignment filtering.
+    sphere: DIPY sphere
+        Sphere for SH to SF projection.
+    exclude_center: bool
+        Whether the center voxel is included in the neighbourhood.
 
     Returns
     -------
-    out: ndarray or float
-        Values at x.
+    weights: ndarray
+        Combined spatial + alignment filter of shape (W, H, D, N) where
+        N is the number of sphere directions.
     """
-    assert sigma > 0.0, "Sigma must be greater than 0."
-    cnorm = 1.0 / sigma / np.sqrt(2.0*np.pi)
-    return cnorm * np.exp(-x**2/2/sigma**2)
+    directions = sphere.vertices.astype(np.float32)
 
+    grid_directions = _get_window_directions(filter_shape).astype(np.float32)
+    distances = np.linalg.norm(grid_directions, axis=-1)
+    grid_directions[distances > 0] = grid_directions[distances > 0] /\
+        distances[distances > 0][..., None]
 
-def _get_window_directions(shape):
+    if sigma_spatial is None:
+        w_spatial = np.ones(filter_shape)
+    else:
+        w_spatial = _evaluate_gaussian_distribution(distances, sigma_spatial)
+
+    if sigma_align is None:
+        w_align = np.ones(np.append(filter_shape, (len(directions),)))
+    else:
+        cos_theta = np.clip(grid_directions.dot(directions.T), -1.0, 1.0)
+        theta = np.arccos(cos_theta)
+        theta[filter_shape[0] // 2,
+              filter_shape[1] // 2,
+              filter_shape[2] // 2] = 0.0
+        w_align = _evaluate_gaussian_distribution(theta, sigma_align)
+
+    # resulting filter
+    w = w_spatial[..., None] * w_align
+
+    if exclude_center:
+        w[filter_shape[0] // 2,
+          filter_shape[1] // 2,
+          filter_shape[2] // 2] = 0.0
+
+    # normalize and return
+    w /= np.sum(w, axis=(0, 1, 2), keepdims=True)
+    return w
+
+
+def _get_sf_range(sh_data, B_mat):
```

Comment:
Arnaud, should this kind of small method go in, say, reconst.utils or something like this? Or, considering that it's only used here, keeping it here with a _ is fine?